third_party/lua-yaml/lyaml.cc

```diff
@@ -819,6 +815,113 @@ static void find_references(struct lua_yaml_dumper *dumper) {
    }
 }
 
+/**
+ * Replace object at `obj_index` on stack with materialized one if it
+ * is found in reftable at `reftable_index` on stack.
+ *
+ * Return true if replace is done and false otherwise.
+ */
+static bool
+materialize_try_reftable(lua_State *L, int obj_index, int reftable_index)
+{
+   assert(reftable_index > 0);
+   if (obj_index < 0)
+      obj_index = lua_gettop(L) + obj_index + 1;
+   lua_pushvalue(L, obj_index);
+   lua_gettable(L, reftable_index);
+   if (!lua_isnil(L, -1)) {
+      /* replace object with materialized one */
+      lua_replace(L, obj_index);
+      return true;
+   }
+   lua_pop(L, 1);
+   return false;
+}
+
+/**
+ * Call luaL_checkfield recursively for object and replace object on stack
+ * with result. Object is at `obj_index` on stack.
+ *
+ * In the process remember the already seen objects in reference table and
+ * check it before materializing next object. Thus the materialized table
+ * keeps the reference structure of the original table. Also the reference
+ * table is checked for luaL_checkfield result so that reference structure
+ * brought by __serialize is kept too. Reference table is at `reftable_index`
+ * on stack.
+ *
+ * Also tables of materialized object keep the hints of __serialize as the
+ * hints affects later serialization.
+ */
+static void
+materialize(struct lua_yaml_dumper *dumper, int obj_index, int reftable_index)
+{
+   struct luaL_field field;
+   lua_State *L = dumper->L;
+
+   assert(reftable_index > 0);
+   if (obj_index < 0)
+      obj_index = lua_gettop(L) + obj_index + 1;
+
+   /* copy original object */
+   lua_pushvalue(L, obj_index);
+   /* check if object is already materialized in reference table */
+   if (materialize_try_reftable(L, -1, reftable_index))
+      goto finish;
+   /* may replace object copy with unpacked value */
+   luaL_checkfield(L, dumper->cfg, -1, &field);
+
+   if (lua_type(L, -1) == LUA_TTABLE) {
+      /* check if table is already materialized in reference table */
+      if (materialize_try_reftable(L, -1, reftable_index))
+         goto finish;
+
+      /* create materialized table */
+      lua_newtable(L);
+      lua_insert(L, -2);
+      int materialized_index = lua_gettop(L) - 1;
+
+      /* save original object -> materialized table in reference table */
+      lua_pushvalue(L, obj_index);
+      lua_pushvalue(L, materialized_index);
+      lua_settable(L, reftable_index);
+      /* save unpacked table -> materialized table in reference table */
+      lua_pushvalue(L, -1);
+      lua_pushvalue(L, materialized_index);
+      lua_settable(L, reftable_index);
+
+      lua_pushnil(L);
+      while (lua_next(L, -2)) {
+         /* copy key for iteration */
+         lua_pushvalue(L, -2);
+         lua_insert(L, -3);
+         materialize(dumper, -1, reftable_index);
+         materialize(dumper, -2, reftable_index);
+         lua_settable(L, materialized_index);
+      }
```

Comment:
Do you consider just creating a reference table (original object -> serialized object mapping) and use it in `find_references()` and `dump_node()` to replace the original table with the serialized one of the fly?

This way we reduce amount of needed temporary GC tracked objects.

Also, it allows the `dump_node()` code to traverse over the original table (if it has no `__serialize`), so we don't need to update so much tests here and in external projects with the same testing approaches.

I've played with it a bit. The patch is [here](https://github.com/Totktonada/tarantool/commit/cd91395d473a2674d6c1f9e69a3b4657db51a92d). It is a little dirty, but at least it seems that this approach is possible to implement and it doesn't look much more complicated.